crates/transaction-emitter-lib/src/transaction_generator/transaction_mix_generator.rs

```diff
@@ -1,72 +1,100 @@
 // Copyright (c) Aptos
 // SPDX-License-Identifier: Apache-2.0
-use crate::transaction_generator::{TransactionGenerator, TransactionGeneratorCreator};
+use crate::{
+    emitter::stats::DynamicStatsTracking,
+    transaction_generator::{TransactionGenerator, TransactionGeneratorCreator},
+};
 use aptos_sdk::types::{transaction::SignedTransaction, LocalAccount};
 use async_trait::async_trait;
 use rand::{prelude::StdRng, Rng};
 use rand_core::{OsRng, SeedableRng};
+use std::sync::Arc;
 
-pub struct TxnMixGenerator {
+pub struct TxnMixPhasedGenerator {
     rng: StdRng,
-    txn_mix: Vec<(Box<dyn TransactionGenerator>, usize)>,
-    total_weight: usize,
+    // for each phase, list of transaction mixes.
+    txn_mix_per_phase: Vec<Vec<(Box<dyn TransactionGenerator>, usize)>>,
+    total_weight_per_phase: Vec<usize>,
+    phase: Arc<DynamicStatsTracking>,
 }
 
-impl TxnMixGenerator {
-    pub fn new(rng: StdRng, txn_mix: Vec<(Box<dyn TransactionGenerator>, usize)>) -> Self {
-        let total_weight = txn_mix.iter().map(|(_, weight)| weight).sum();
+impl TxnMixPhasedGenerator {
+    pub fn new(
+        rng: StdRng,
+        txn_mix_per_phase: Vec<Vec<(Box<dyn TransactionGenerator>, usize)>>,
+        phase: Arc<DynamicStatsTracking>,
+    ) -> Self {
+        let total_weight_per_phase = txn_mix_per_phase
+            .iter()
+            .map(|txn_mix| txn_mix.iter().map(|(_, weight)| weight).sum())
+            .collect();
         Self {
             rng,
-            txn_mix,
-            total_weight,
+            txn_mix_per_phase,
+            total_weight_per_phase,
+            phase,
         }
     }
 }
 
-impl TransactionGenerator for TxnMixGenerator {
+impl TransactionGenerator for TxnMixPhasedGenerator {
     fn generate_transactions(
         &mut self,
         accounts: Vec<&mut LocalAccount>,
         transactions_per_account: usize,
     ) -> Vec<SignedTransaction> {
-        let mut picked = self.rng.gen_range(0, self.total_weight);
-        for (gen, weight) in &mut self.txn_mix {
+        let phase = self.phase.get_cur_phase() % self.txn_mix_per_phase.len();
```

Comment:
Why `%` is needed? It should be `let phase = self.phase.get_cur_phase()` ?

Reply:
I allow for simplicity to pass a single phase, and always use that one. I can be more explicit here. (for example if there is warmup, test and cooldown phase, no need to specify all 3